horovod/common/controller.cc

```diff
@@ -0,0 +1,735 @@
+  }
+
+  // Copy the data structures out from parameters.
+  // However, don't keep the lock for the rest of the loop, so that
+  // enqueued stream callbacks can continue.
+
+  CacheCoordinator cache_coordinator(response_cache_.num_active_bits());
+
+  // message queue used only in this cycle
+  std::deque<Request> message_queue_tmp;
+  tensor_queue_.PopMessagesFromQueue(message_queue_tmp);
+  for (auto& message : message_queue_tmp) {
+    // Keep track of cache hits
+    if (response_cache_.capacity() > 0) {
+      auto cache_ = response_cache_.cached(message);
+      if (cache_ == ResponseCache::CacheState::HIT) {
+        uint32_t cache_bit = response_cache_.peek_cache_bit(message);
+        cache_coordinator.record_hit(cache_bit);
+
+        // Record initial time cached tensor is encountered in queue.
+        stall_inspector_.RecordCacheTensorStart(message.tensor_name());
+
+      } else {
+        if (cache_ == ResponseCache::CacheState::INVALID) {
+          uint32_t cache_bit = response_cache_.peek_cache_bit(message);
+          cache_coordinator.record_invalid_bit(cache_bit);
+        }
+        cache_coordinator.set_uncached_in_queue(true);
+
+        // Remove timing entry if uncached or marked invalid.
+        stall_inspector_.RemoveCachedTensor(message.tensor_name());
+      }
+    }
+  }
+
+  // Flag indicating that the background thread should shut down.
+  bool should_shut_down = shut_down_;
+
+  // Check for stalled tensors.
+  if (stall_inspector_.ShouldPerformCheck()) {
+    if (is_coordinator_) {
+      should_shut_down |= stall_inspector_.CheckForStalledTensors(size_);
+    }
+
+    if (response_cache_.capacity() > 0) {
+      stall_inspector_.InvalidateStalledCachedTensors(cache_coordinator,
+                                                      response_cache_);
+    }
+    stall_inspector_.UpdateCheckTime();
+  }
+
+  cache_coordinator.set_should_shut_down(should_shut_down);
+
+  if (response_cache_.capacity() > 0) {
+    // Obtain common cache hits and cache invalidations across workers. Also,
+    // determine if any worker has uncached messages in queue or requests
+    // a shutdown. This function removes any invalid cache entries, if they
+    // exist.
+    CoordinateCacheAndState(cache_coordinator);
+
+    // Remove uncommon cached tensors from queue and replace to state
+    // queue for next cycle. Skip adding common cached tensors to
+    // queue as they are handled separately.
+    size_t num_messages = message_queue_tmp.size();
+    for (size_t i = 0; i < num_messages; ++i) {
+      auto message = message_queue_tmp.front();
+      if (response_cache_.cached(message) == ResponseCache::CacheState::HIT) {
+        uint32_t cache_bit = response_cache_.peek_cache_bit(message);
+        if (cache_coordinator.cache_hits().find(cache_bit) ==
+            cache_coordinator.cache_hits().end()) {
+          // Try to process again in next cycle.
+          tensor_queue_.PushMessageToQueue(std::move(message));
```

Comment:
This will cause a lot of repeated locks for big models like resnet152. Previously, we locked one time here. We should profile if this causes any performance degradation.

Reply:
That's a good thought, but the benchmark shows little difference actually. Maybe it's because we are fusing the most tensors.